L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,905 @@
+      int waferu = 0;
+      int waferv = 0;
+      unpackWaferId(wafer_itr->second, waferu, waferv);
+      DetId::Detector det = (module_si_id.subdetId() == ForwardSubdetector::HGCEE ? DetId::HGCalEE : DetId::HGCalHSi);
+      HGCalTriggerSubdetector subdet =
+          (module_si_id.subdetId() == ForwardSubdetector::HGCEE ? HGCalTriggerSubdetector::HGCalEETrigger
+                                                                : HGCalTriggerSubdetector::HGCalHSiTrigger);
+      unsigned layer = module_si_id.layer();
+      unsigned wafer_type = detIdWaferType(det, layer, waferu, waferv);
+      int nroc = (wafer_type == HGCSiliconDetId::HGCalFine ? 6 : 3);
+      // Loop on ROCs in wafer
+      for (int roc = 1; roc <= nroc; roc++) {
+        // loop on TCs in ROC
+        auto tc_uvs = tc2roc.getTriggerId(roc, wafer_type);
+        for (const auto& tc_uv : tc_uvs) {
+          HGCalTriggerDetId trigger_cell_id(
+              subdet, module_si_id.zside(), wafer_type, layer, waferu, waferv, tc_uv.first, tc_uv.second);
+          trigger_cell_det_ids.emplace(trigger_cell_id);
+        }
+      }
+    }
+  }
+  return trigger_cell_det_ids;
+}
+
+HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp3::getNeighborsFromTriggerCell(
+    const unsigned trigger_cell_id) const {
+  throw cms::Exception("FeatureNotImplemented") << "Neighbor search is not implemented in HGCalTriggerGeometryV9Imp3";
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::getLinksInModule(const unsigned module_id) const {
+  DetId module_det_id(module_id);
+  unsigned links = 0;
+  // Scintillator
+  if (module_det_id.det() == DetId::HGCalHSc) {
+    links = hSc_links_per_module_;
+  } else if (module_det_id.det() == DetId::Forward && module_det_id.subdetId() == ForwardSubdetector::HFNose) {
+    links = 1;
+  }
+  // TO ADD HFNOSE : getLinksInModule
+  // Silicon
+  else {
+    HGCalDetId module_det_id_si(module_id);
+    unsigned module = module_det_id_si.wafer();
+    unsigned layer = layerWithOffset(module_id);
+    module = (module & sector0_mask_);
+    links = links_per_module_.at(packLayerModuleId(layer, module));
+  }
+  return links;
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::getModuleSize(const unsigned module_id) const {
+  DetId module_det_id(module_id);
+  unsigned nWafers = 1;
+  // Scintillator
+  if (module_det_id.det() == DetId::HGCalHSc) {
+    nWafers = hSc_wafers_per_module_;
+  }
+  // Check for HFNOSE : getModuleSize
+  // Silicon
+  else {
+    HGCalDetId module_det_id_si(module_id);
+    unsigned module = module_det_id_si.wafer();
+    unsigned layer = layerWithOffset(module_id);
+    nWafers = module_to_wafers_.count(packLayerModuleId(layer, module));
+  }
+  return nWafers;
+}
+
+std::vector<unsigned> HGCalTriggerGeometryV9Imp3::getStage1FpgasFromStage2Fpga(const unsigned stage2_id) const {
+
+  std::vector<unsigned> stage1_ids = mapping_config_["Stage2"][stage2_id]["Stage1"];
```

Comment:
The mapping information should be stored into `unordered_map` or `unordered_multimap` for fast access. The translation from the json format to maps should be done only in the initialization.
Geometry accesses are among the most CPU consuming in the HGC TPG, so these mapping accesses should be as fast as possible. Unordered maps with integer keys should be used because of their constant access complexity.